### PR TITLE
Google OAuth2 interface (only for FYKOS orgs)

### PR DESCRIPTION
**Pull request by cevro**

### CODE REVIEW

File: app/Modules/CoreModule/presenters/AuthenticationPresenter.php [hunk 1]
```diff
@@ -58,17 +63,21 @@ final class AuthenticationPresenter extends BasePresenter {
     private IGlobalSession $globalSession;
     private PasswordAuthenticator $passwordAuthenticator;
     private AccountManager $accountManager;
+    private Google $provider;
```
Comment (Werkov) [inline app/Modules/CoreModule/presenters/AuthenticationPresenter.php:66]: More descriptive name, perhaps `$googleProvider`? But see below my note about it.

File: app/Modules/CoreModule/presenters/AuthenticationPresenter.php [hunk 2]
```diff
@@ -331,4 +375,25 @@ protected function beforeRender(): void {
...
+
+    public function getGoogleSection(): SessionSection {
+        return $this->getSession()->getSection('google-oauth2state');
+    }
+
+    /**
+     * @return Google
+     * @throws InvalidLinkException
+     */
+    protected function getGoogleOAuth2Provider(): Google {
```
Comment (Werkov) [inline app/Modules/CoreModule/presenters/AuthenticationPresenter.php:387]: Wouldn't it make more sense to inject the whole `Google` class instance via standard dependency injection and defined in `services.neon`?

File: app/model/Authentication/GoogleAuthenticator.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+use FKSDB\ORM\Services\ServicePersonInfo;
+use FKSDB\YearCalculator;
+use Nette\Database\Table\ActiveRow;
+use Nette\Security\AuthenticationException;
+use Tracy\Debugger;
+
+/**
+ * Due to author's laziness there's no class doc (or it's self explaining).
+ *
+ * @author Michal Koutný <michal@fykos.cz>
```
Comment (Werkov) [inline app/model/Authentication/GoogleAuthenticator.php:22]: Please, don't copy-paste my authorship to newly created classes.

File: app/model/Authentication/GoogleAuthenticator.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * @param array $user
+     * @return ModelPerson|null
+     * @throws AuthenticationException
+     * @throws InvalidCredentialsException
+     */
+    private function findPerson(array $user): ?ModelPerson {
+        if (!$user['email']) {
+            throw new AuthenticationException(_('V profilu google nebyl nalezen e-mail.'));
```
Comment (Werkov) [inline app/model/Authentication/GoogleAuthenticator.php:68]: s/google/Google/

File: app/model/Authentication/GoogleAuthenticator.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                break;
+            default:
+                throw new InvalidCredentialsException();
+        }
+        /** @var ModelOrg|null $org */
+        $org = $this->serviceOrg->getTable()->where(['domain_alias' => $domainAlias, 'contest_id' => $contestId])->fetch();
+        return $org ? $org->getPerson() : null;
+
+        /* $result = $this->serviceOrg->getTable()->where(':person_info.email = ?', $user['email'])->fetch();
+         return $result;*/
```
Comment (Werkov) [inline app/model/Authentication/GoogleAuthenticator.php:86]: Drop this.

File: app/model/WebService/WebServiceModel.php [hunk 6]
```diff
@@ -31,28 +32,20 @@ class WebServiceModel {
...
-
     private StatsModelFactory $statsModelFactory;
-
-    /** @var ModelLogin */
-    private $authenticatedLogin;
-
-    private IAuthenticator $authenticator;
-
+    private ModelLogin $authenticatedLogin;
+    private PasswordAuthenticator $authenticator;
```
Comment (Werkov) [inline app/model/WebService/WebServiceModel.php:39]: Good change.

File: app/model/Authentication/GoogleAuthenticator.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+     */
+    public function authenticate(array $user): ModelLogin {
+        $person = $this->findPerson($user);
+
+        if (!$person) {
+            throw new UnknownLoginException();
+        } else {
+            $login = $person->getLogin();
+            if (!$login) {
+                $login = $this->accountManager->createLogin($person);
```
Comment (Werkov) [inline app/model/Authentication/GoogleAuthenticator.php:50]: Is it intended that org-level accounts are created automatically?

File: app/Modules/CoreModule/templates/Authentication/login.latte [hunk 8]
```diff
@@ -8,5 +8,6 @@
         {input send class => 'btn btn-lg btn-success btn-block'}
         <a n:href=":Public:Register:default" class="btn btn-info btn-block">{_'Registrace'}</a>
         <a class="btn btn-secondary btn-sm btn-block" n:href="recover, backlink => null, bc => null, flag => null">{_'Zapomenuté heslo'}</a>
+        <a class="btn btn-secondary btn-sm btn-block" href="{link google!}"><i class="fa fa-google mr-3"/>{_'Login via FYKOS account on Google'}</a>
```
Comment (Werkov) [inline app/Modules/CoreModule/templates/Authentication/login.latte:11]: Are the canonical labels in English or Czech?

File: app/model/Authentication/GoogleAuthenticator.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    private function findPerson(array $user): ?ModelPerson {
+        if (!$user['email']) {
+            throw new AuthenticationException(_('V profilu google nebyl nalezen e-mail.'));
+        }
+        [$domainAlias, $domain] = explode('@', $user['email']);
+        switch ($domain) {
+            case 'fykos.cz':
+                $contestId = ModelContest::ID_FYKOS;
+                break;
+            case 'vyfuk.org':
```
Comment (Werkov) [inline app/model/Authentication/GoogleAuthenticator.php:75]: This domain doesn't seem to be handled by G Suite.

**Review by Werkov — CHANGES_REQUESTED** [review]
Some more and less relevant notes are in the code.
Two general questions:
- Is FacebookAuthenticator still working? Is it used?
- How is ensured that only FYKOS organization accounts can be used to gain "org" access (and not anyone with their Google account)?

File: app/Modules/CoreModule/presenters/AuthenticationPresenter.php (hunk 2 above)
Comment (cevro) [reply app/Modules/CoreModule/presenters/AuthenticationPresenter.php:387]: Nice idea, but in constructor Google(provider) you require redirect link, that's mean, you need presenter to create instance of Google(provider). Of course you can create `Factory` with method `create(IPresenter):Google`, but it's IMO over-kill, cause  google is used only in single Presenter.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 3 above)
Comment (cevro) [reply app/model/Authentication/GoogleAuthenticator.php:22]: Sry, I fix it,
...but core of this factory is your own code.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 5 above)
Comment (cevro) [reply app/model/Authentication/GoogleAuthenticator.php:86]: Its prepare for using google outside of FYKOS google accounts.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 7 above)
Comment (cevro) [reply app/model/Authentication/GoogleAuthenticator.php:50]: Its only record in table `login`, records in tables `person*` and `org` will be created as before, accManager should not create it.

File: app/Modules/CoreModule/templates/Authentication/login.latte (hunk 8 above)
Comment (cevro) [reply app/Modules/CoreModule/templates/Authentication/login.latte:11]: Canonical is English, but it's still many places where isn't.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 9 above)
Comment (cevro) [reply app/model/Authentication/GoogleAuthenticator.php:75]: Answer in email.

**Comment by cevro** [comment]
@Werkov 

> Is FacebookAuthenticator still working? Is it used?
No, not working approx 4-5 years, expire FB SDK, and anybody didn't fix it, but code is still in FKSDB and probably is still registered as service.

> How is ensured that only FYKOS organization accounts can be used to gain "org" access (and not anyone with their Google account)?
App ID is registered only for FYKOS's G Suite, and it's possible restrict users only for domain. If will works fine, I remove restriction and allow login for all google users.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 7 above)
Comment (Werkov) [reply app/model/Authentication/GoogleAuthenticator.php:49]: Fair enough.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 5 above)
Comment (Werkov) [reply app/model/Authentication/GoogleAuthenticator.php]: I'd suggest keeping it in a (future) separate commit, without the comment in the current code.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 3 above)
Comment (Werkov) [reply app/model/Authentication/GoogleAuthenticator.php]: I believe it's just the skeleton, the meat is yours. Feel free to remove my attribution.

File: app/Modules/CoreModule/presenters/AuthenticationPresenter.php (hunk 2 above)
Comment (Werkov) [reply app/Modules/CoreModule/presenters/AuthenticationPresenter.php:387]: I missed the link dependency. What about https://api.nette.org/2.4/Nette.Application.LinkGenerator.html?

**Comment by Werkov** [comment]
Thanks for the explanation. Ad FB authenticator, I'd suggest delete it (with roots) so that there is less dead/vulnerable code in the auth area.

File: app/model/Authentication/GoogleAuthenticator.php (hunk 3 above)
Comment (cevro) [reply app/model/Authentication/GoogleAuthenticator.php]: Well, it's done.

File: app/Modules/CoreModule/presenters/AuthenticationPresenter.php (hunk 2 above)
Comment (cevro) [reply app/Modules/CoreModule/presenters/AuthenticationPresenter.php]: Good point, I use LinkGenerator in some factories, there are some positives/negatives link has to be always absolute and, and starting ':' has to be omitted. There is a big incompatibility between links in Presenters and LinkGenerator. But I tried it, and looks fine.

File: app/model/Authentication/GoogleAuthenticator.php [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+                $contestId = ModelContest::ID_VYFUK;
+                break;
+            default:
+                throw new InvalidCredentialsException();
```
Comment (sstenchlak) [inline app/model/Authentication/GoogleAuthenticator.php:73]: I would prefer something like 'This e-mail address is not linked to an **organizer** account' exception to emphasize that this option is not for contestants.

**Review by sstenchlak — APPROVED** [review]

**Review by sstenchlak — APPROVED** [review]

**Pull request merged**